File: clients/typescript/src/identity.ts
import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign,
  type KeyObject,
} from "node:crypto";

import type { IntegrationGrant } from "./generated/protocol.js";
import { RuntimeProtocolError } from "./errors.js";

const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export class IntegrationIdentity {
  readonly #privateKey: KeyObject;

  private constructor(privateKey: KeyObject) {
    if (privateKey.asymmetricKeyType !== "ed25519") {
      throw new RuntimeProtocolError("integration identity is not an Ed25519 private key");
    }
    this.#privateKey = privateKey;
  }

  public static generate(): IntegrationIdentity {
    const { privateKey } = generateKeyPairSync("ed25519");
    return new IntegrationIdentity(privateKey);
  }

  public static fromPkcs8(bytes: Uint8Array): IntegrationIdentity {
    try {
      return new IntegrationIdentity(createPrivateKey({
        key: Buffer.from(bytes),
        format: "der",
        type: "pkcs8",
      }));
    } catch (error) {
      throw new RuntimeProtocolError(`integration private key is malformed: ${String(error)}`);
    }
  }

  public exportPkcs8(): Uint8Array {
    return this.#privateKey.export({ format: "der", type: "pkcs8" });
  }

  public publicKeyBase64(): string {
    const encoded = createPublicKey(this.#privateKey).export({ format: "der", type: "spki" });
    if (encoded.byteLength !== ED25519_SPKI_PREFIX.byteLength + 32
      || !encoded.subarray(0, ED25519_SPKI_PREFIX.byteLength).equals(ED25519_SPKI_PREFIX)) {
      throw new RuntimeProtocolError("integration public key has an unexpected Ed25519 encoding");
    }
    return encoded.subarray(ED25519_SPKI_PREFIX.byteLength).toString("base64url");
  }

  public signBase64(payload: Uint8Array): string {
    return sign(null, payload, this.#privateKey).toString("base64url");
  }
}

export class IntegrationCredentials {
  public constructor(
    public readonly identity: IntegrationIdentity,
    public readonly grant: IntegrationGrant,
  ) {}
}

File: crates/runtrol-runtime-client/src/identity.rs
//! Consumer-owned signing identity and approved reconnect credentials.

use base64ct::{Base64UrlUnpadded, Encoding as _};
use ed25519_dalek::{Signer as _, SigningKey};
use runtrol_runtime_protocol::{
    IntegrationGrant, PendingEnrollmentId, self_approval_signing_payload,
};

use crate::ClientError;

/// One consumer-owned Ed25519 identity. Runtime receives only its public key.
#[derive(Clone)]
pub struct IntegrationIdentity {
    signing: SigningKey,
}

impl core::fmt::Debug for IntegrationIdentity {
    fn fmt(&self, formatter: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        formatter.write_str("IntegrationIdentity([redacted])")
    }
}

impl IntegrationIdentity {
    /// Generate a new installed-integration identity.
    ///
    /// # Errors
    ///
    /// The operating system random source was unavailable.
    pub fn generate() -> Result<Self, ClientError> {
        let mut secret = [0_u8; 32];
        getrandom::fill(&mut secret).map_err(|error| {
            ClientError::Protocol(format!(
                "the integration identity could not be generated: {error}"
            ))
        })?;
        Ok(Self::from_secret_bytes(secret))
    }

    /// Restore an identity from consumer-owned secure storage.
    #[must_use]
    pub fn from_secret_bytes(secret: [u8; 32]) -> Self {
        Self {
            signing: SigningKey::from_bytes(&secret),
        }
    }

    /// Export bytes for the consumer's own operating-system secure storage.
    #[must_use]
    pub fn secret_bytes(&self) -> [u8; 32] {
        self.signing.to_bytes()
    }

    /// Base64url public verification key used in an enrollment manifest.
    #[must_use]
    pub fn public_key_base64(&self) -> String {
        Base64UrlUnpadded::encode_string(self.signing.verifying_key().as_bytes())
    }

    /// Sign the one pending enrollment identity for a local self-approval request.
    ///
    /// This deliberately exposes only the domain-separated self-approval proof rather than a general signing
    /// primitive. The returned signature cannot authenticate a Runtime connection or another enrollment.
    ///
    /// # Errors
    ///
    /// Canonical protocol serialization failed, which indicates an implementation defect.
    pub fn self_approval_signature(
        &self,
        pending_id: &PendingEnrollmentId,
    ) -> Result<String, ClientError> {
        let payload = self_approval_signing_payload(pending_id).map_err(|error| {
            ClientError::Protocol(format!(
                "the self-approval proof payload cannot be encoded: {error}"
            ))
        })?;
        Ok(self.sign_base64(&payload))
    }

    pub(crate) fn sign_base64(&self, payload: &[u8]) -> String {
        Base64UrlUnpadded::encode_string(&self.signing.sign(payload).to_bytes())
    }
}

/// Consumer-owned identity paired with the current approved public grant.
#[derive(Clone, Debug)]
pub struct IntegrationCredentials {
    identity: IntegrationIdentity,
    grant: IntegrationGrant,
}

impl IntegrationCredentials {
    /// Bind the locally stored private identity to a grant returned for that identity.
    #[must_use]
    pub const fn new(identity: IntegrationIdentity, grant: IntegrationGrant) -> Self {
        Self { identity, grant }
    }

    /// Consumer-owned signing identity.
    #[must_use]
    pub const fn identity(&self) -> &IntegrationIdentity {
        &self.identity
    }

    /// Current reconnect grant and generations.
    #[must_use]
    pub const fn grant(&self) -> &IntegrationGrant {
        &self.grant
    }

    pub(crate) fn into_parts(self) -> (IntegrationIdentity, IntegrationGrant) {
        (self.identity, self.grant)
    }
}
